### ISSUE TITLE
command line interface

### ISSUE THREAD

**Issue by karubn** [Anjok07/ultimatevocalremovergui#1288]
uvr5 .ok it is a nice program, we are expecting command line feature as soon as possible, thank you in advance. anjok07

**Comment by karubn** [comment]
The uvr gui version is nice and successful, but we are patiently waiting for the windows command line interface to be a little practical.

**Comment by brianyuastri** [comment]
from gui_data.constants import VR_ARCH_TYPE, MDX_ARCH_TYPE, DEMUCS_ARCH_TYPE
from separate import SeperateDemucs, SeperateMDX, SeperateMDXC, SeperateVR
from UVR import ModelData

audio_file = 'your_audio_file'
export_path = 'your_export_folder'

arch_type = 'MDX-Net'
model_name = 'MDX23C-InstVoc HQ'

current_model = ModelData(model_name, arch_type)
current_model.mdxnet_stem_select = 'Vocals'
audio_file_base = audio_file.split('/')[-1].split('.')[0]
def do_nothing(*args, **kwargs): pass
def return_none(*args, **kwargs): return None, None
all_models = [model_name]

process_data = {
    'model_data': current_model, 
    'export_path': export_path,
    'audio_file_base': audio_file_base,
    'audio_file': audio_file,
    'set_progress_bar': do_nothing,
    'write_to_console': do_nothing,
    'process_iteration': do_nothing,
    'cached_source_callback': return_none,
    'cached_model_source_holder': do_nothing,
    'list_all_models': all_models,
    'is_ensemble_master': False,
    'is_4_stem_ensemble': False
}

if current_model.process_method == VR_ARCH_TYPE:
    seperator = SeperateVR(current_model, process_data)
if current_model.process_method == MDX_ARCH_TYPE:
    seperator = SeperateMDXC(current_model, process_data) if current_model.is_mdx_c else SeperateMDX(current_model, process_data)
if current_model.process_method == DEMUCS_ARCH_TYPE:
    seperator = SeperateDemucs(current_model, process_data)

seperator.seperate()

**Comment by karubn** [comment]
I think this is python code, isn't it? I need something with a windows command line interface or a batch file.

**Comment by brianyuastri** [comment]
does

python your_script.py

not work in the command line? I am not sure what you need here

**Comment by karubn** [comment]
The code above is Python code. I think I need something like a Windows command line interface or a batch file. As far as I understand, if the developer kindly requests it, we are waiting for Windows CLI.

**Comment by brianyuastri** [comment]
Since you keep insisting that the Python code has nothing to do with CLI, I assume that you are unfamiliar with Python? Otherwise, you can visit

https://docs.python.org/3/library/argparse.html

and apply it to the code to suit your needs.

**Comment by karubn** [comment]
Yes, unfortunately, I am not familiar with Python. We are patiently waiting for Windows CLI.

**Comment by ehsanbina** [comment]
@brianyuastri

Hi,
In which directory of UVR should we create file from this python code and run the file so that the imports are done correctly?
I created the file in the root folder of the software and ran it, but it got an error from importing the module.
C:\Users\username\AppData\Local\Programs\Ultimate Vocal Remover

![Screenshot 2024-08-29 162604](https://github.com/user-attachments/assets/583a4402-cf29-4106-8bb5-5900cfd21b91)

Please advise more
thanks

**Comment by tamirpariente** [comment]
> from gui_data.constants import VR_ARCH_TYPE, MDX_ARCH_TYPE, DEMUCS_ARCH_TYPE from separate import SeperateDemucs, SeperateMDX, SeperateMDXC, SeperateVR from UVR import ModelData
> 
> audio_file = 'your_audio_file' export_path = 'your_export_folder'
> 
> arch_type = 'MDX-Net' model_name = 'MDX23C-InstVoc HQ'
> 
> current_model = ModelData(model_name, arch_type) current_model.mdxnet_stem_select = 'Vocals' audio_file_base = audio_file.split('/')[-1].split('.')[0] def do_nothing(*args, **kwargs): pass def return_none(*args, **kwargs): return None, None all_models = [model_name]
> 
> process_data = { 'model_data': current_model, 'export_path': export_path, 'audio_file_base': audio_file_base, 'audio_file': audio_file, 'set_progress_bar': do_nothing, 'write_to_console': do_nothing, 'process_iteration': do_nothing, 'cached_source_callback': return_none, 'cached_model_source_holder': do_nothing, 'list_all_models': all_models, 'is_ensemble_master': False, 'is_4_stem_ensemble': False }
> 
> if current_model.process_method == VR_ARCH_TYPE: seperator = SeperateVR(current_model, process_data) if current_model.process_method == MDX_ARCH_TYPE: seperator = SeperateMDXC(current_model, process_data) if current_model.is_mdx_c else SeperateMDX(current_model, process_data) if current_model.process_method == DEMUCS_ARCH_TYPE: seperator = SeperateDemucs(current_model, process_data)
> 
> seperator.seperate()

Did you actually tested this script?
Because i tried it after I already tried many times to change the code to fit to a single short python script, but the thing is that you have this "root" (the gui itself) used everywhere.. :(

**Comment by chameleon-ai** [comment]
After trying to get this to work, I can see why it's not officially supported. The main problem is that the ModelData class, which separate() relies on, is tightly integrated with the GUI itself. I had to copy ModelData to a new module and hack away all the root dependencies. Unfortunately, in my haste to just get it to run, I only got it to work with one specific MDX_NET model. Pretty sure VX, Ensemble mode, Denoise, Karaoke, etc. won't work. But it does run inference on the GPU, or CPU if you like.

I wrote this assuming the script is a directory up from the ultimatevocalremovergui directory, but it should easily run in the same directory as UVR.py if you just remove the ultimatevocalremovergui prefixes at the top of the file:
[uvr_cli.py.txt](https://github.com/user-attachments/files/19261293/uvr_cli.py.txt)
(no guarantees, it's very rough, but it works on my machine: Linux, python 3.10, pytorch ROCM)

Interestingly, separate.py has a smaller subset of requirements than UVR.py:
[requirements.txt](https://github.com/user-attachments/files/19261296/requirements.txt)
